File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(
  items: T[],
  predicate: (item: T) => boolean,
  consumer: (item: T) => U
): U | undefined {
  const foundItem = items.find(predicate)
  return foundItem !== undefined ? consumer(foundItem) : undefined
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  let power = 1n // Start with BigInt 1
  while (true) {
    yield power
    power *= base // Multiply using BigInt
  }
}

export async function meaningfulLineCount(filePath: string): Promise<number> {
  const fileHandle = await open(filePath, "r") // Open the file for reading
  try {
    const fileContent = await fileHandle.readFile({ encoding: "utf-8" })
    const lines = fileContent.split(/\r?\n/) // Split content into lines (handles both Unix and Windows line breaks)

    let lineCount = 0

    for (const line of lines) {
      const trimmedLine = line.trim()

      if (trimmedLine.length === 0 || trimmedLine.startsWith("#")) {
        continue // Skip empty lines or lines starting with #
      }

      lineCount++
    }

    return lineCount
  } finally {
    await fileHandle.close() // Ensure the file is closed
  }
}

interface Sphere {
  kind: "Sphere"
  radius: number
}

interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2
    case "Box":
      return (
        2 *
        (shape.width * shape.length +
          shape.width * shape.depth +
          shape.length * shape.depth)
      )
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * shape.radius ** 3
    case "Box":
      return shape.width * shape.length * shape.depth
  }
}

// Custom toString function for both Sphere and Box
export function toString(shape: Shape): string {
  switch (shape.kind) {
    case "Sphere":
      return `Sphere(radius: ${shape.radius})`
    case "Box":
      return `Box(width: ${shape.width}, height: ${shape.length}, depth: ${shape.depth})`
  }
}

export interface BinarySearchTree<T> {
  size(): number
  insert(value: T): BinarySearchTree<T>
  contains(value: T): boolean
  inorder(): Iterable<T>
}

class Node<T> implements BinarySearchTree<T> {
  constructor(
    public value: T,
    public left: BinarySearchTree<T>,
    public right: BinarySearchTree<T>
  ) {}

  size(): number {
    return 1 + this.left.size() + this.right.size()
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      const newLeft = this.left.insert(value)
      return new Node(this.value, newLeft, this.right)
    } else if (value > this.value) {
      const newRight = this.right.insert(value)
      return new Node(this.value, this.left, newRight)
    }
    // If value is equal, do not insert and return this (immutable)
    return this // Ensuring immutability
  }

  contains(value: T): boolean {
    if (value === this.value) {
      return true
    } else if (value < this.value) {
      return this.left.contains(value)
    } else {
      return this.right.contains(value)
    }
  }

  *inorder(): Iterable<T> {
    yield* this.left.inorder()
    yield this.value
    yield* this.right.inorder()
  }

  toString(): string {
    const leftStr = this.left.toString()
    const rightStr = this.right.toString()
    const leftPart = leftStr ? `(${leftStr})` : ""
    const rightPart = rightStr ? `(${rightStr})` : ""
    return `${leftPart}${this.value}${rightPart}`
  }
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty(), new Empty())
  }

  contains(value: T): boolean {
    return false
  }

  *inorder(): Iterable<T> {
    return [] // Empty tree yields no values
  }

  toString(): string {
    return "()" // Empty tree representation
  }
}

File: haskell/Exercises.hs
module Exercises
    ( change,
    -- put the proper exports here
    firstThenApply,
    powers,
    meaningfulLineCount,
    volume,
    surfaceArea, 
    Shape(Box,Sphere),
    BST(Empty),
    size,
    insert,
    inorder,
    contains
    ) where

import qualified Data.Map as Map
import Data.Text (pack, unpack, replace)
import Data.List(isPrefixOf, find)
import Data.Char(isSpace)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
    | amount < 0 = Left "amount cannot be negative"
    | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
        where
          changeHelper [] remaining counts = counts
          changeHelper (d:ds) remaining counts =
            changeHelper ds newRemaining newCounts
              where
                (count, newRemaining) = remaining `divMod` d
                newCounts = Map.insert d count counts

-- Write your first then apply function here
firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs p f = fmap f (find p xs)
  where
    find _ [] = Nothing
    find p (y:ys)
      | p y       = Just y
      | otherwise = find p ys


-- Write your infinite powers generator here
powers :: (Integral a) => a -> [a]
powers base = map (base^) [0..]

-- Write your line count function here
meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount filePath = do 
  document <- readFile filePath
  let allWiteSpace = all isSpace
      trimStart = dropWhile isSpace
      isMeaningful line = 
        not (allWiteSpace line) &&
        not ("#" `isPrefixOf` (trimStart line))
  return $ length $ filter isMeaningful $ lines document
  
-- Write your shape data type here
data Shape 
  = Sphere Double 
  | Box Double Double Double
   deriving (Eq,Show);

volume :: Shape -> Double  
volume (Sphere r) = (4 / 3) * pi * r^3 
volume (Box l w h) = l * w * h  

   -- do the same for surface area
surfaceArea :: Shape -> Double
surfaceArea (Sphere r) = 4  * pi * r^2
surfaceArea (Box l w h) = ((2*l * w)+(2*l * h)+(2*h* w))


-- Write your binary search tree algebraic type here
data BST a 
  = Empty 
  | Node a (BST a) (BST a)

size :: BST a -> Int
size Empty = 0
size (Node _ left right) = 1 + size left + size right

contains :: Ord a => a -> BST a -> Bool
contains _ Empty = False
contains value (Node nodeValue left right)
  | value == nodeValue = True
  | value < nodeValue  = contains value left
  | otherwise          = contains value right

inorder :: BST a -> [a]
inorder Empty = []
inorder (Node value left right) = inorder left ++ [value] ++ inorder right

insert :: Ord a => a -> BST a -> BST a
insert value Empty = Node value Empty Empty
insert value (Node nodeValue left right)
    | value < nodeValue = Node nodeValue (insert value left) right
    | value > nodeValue = Node nodeValue left (insert value right)
    | otherwise = Node nodeValue left right  

instance Show a => Show (BST a) where
    show Empty = "()"
    show (Node value Empty Empty) = "(" ++ show value ++ ")"  
    show (Node value left right) =
        let leftStr = show left
            rightStr = show right
        in case (leftStr, rightStr) of
            ("()", "()") -> "(" ++ show value ++ ")"
            ("()", _)    -> "(" ++ show value ++ rightStr ++ ")"
            (_, "()")    -> "(" ++ leftStr ++ show value ++ ")"
            _            -> "(" ++ leftStr ++ show value ++ rightStr ++ ")"






